src/executor/evaluate/evaluated.rs

```diff
@@ -182,126 +146,145 @@ impl TryInto<Value> for Evaluated<'_> {
-
-        let subtract_value = |l: &data::Value, other: &Evaluated<'a>| match other {
-            LiteralRef(r) => l.subtract(&l.clone_by(r)?).map(Evaluated::Value),
-            Literal(r) => l.subtract(&l.clone_by(&r)?).map(Evaluated::Value),
-            ValueRef(r) => l.subtract(r).map(Evaluated::Value),
-            Value(r) => l.subtract(&r).map(Evaluated::Value),
-            StringRef(_) => unreachable(),
-        };
+    };
+}
+macro_rules! unary_op {
+    ($name:ident, $op:tt) => {
+        pub fn $name(&self) -> Result<Evaluated<'a>> {
+            let literal_unary_op = |v: &&AstValue| match v {
+                AstValue::Number(v) => v
+                    .parse::<i64>()
+                    .map_or_else(
+                        |_| v.parse::<f64>().map(|v| AstValue::Number((0.0 $op v).to_string())),
+                        |v| Ok(AstValue::Number((0 $op v).to_string())),
+                    )
+                    .map_err(|_| EvaluateError::LiteralUnaryOperationOnNonNumeric.into()),
+                AstValue::Null => Ok(AstValue::Null),
+                _ => Err(EvaluateError::LiteralUnaryOperationOnNonNumeric.into()),
+            };
 
-        match self {
-            LiteralRef(l) => subtract_literal(l, other),
-            Literal(l) => subtract_literal(&l, other),
-            ValueRef(l) => subtract_value(l, other),
-            Value(l) => subtract_value(&l, other),
-            StringRef(_) => unreachable(),
+            match self {
+                LiteralRef(v) => literal_unary_op(v).map(Evaluated::Literal),
+                Literal(v) => literal_unary_op(&v).map(Evaluated::Literal),
+                ValueRef(v) => v.$name().map(Evaluated::Value),
+                Value(v) => (&v).$name().map(Evaluated::Value),
+                _ => Err(EvaluateError::UnreachableEvaluatedArithmetic.into()),
+            }
         }
-    }
-
-    pub fn multiply(&self, other: &Evaluated<'a>) -> Result<Evaluated<'a>> {
-        use Evaluated::*;
-
-        let unreachable = || Err(EvaluateError::UnreachableEvaluatedArithmetic.into());
-
-        let multiply_literal = |l, other: &Evaluated<'a>| match other {
-            LiteralRef(r) => literal_multiply(l, r).map(Evaluated::Literal),
-            Literal(r) => literal_multiply(l, &r).map(Evaluated::Literal),
-            ValueRef(r) => (r.clone_by(l)?).multiply(r).map(Evaluated::Value),
-            Value(r) => (r.clone_by(l)?).multiply(r).map(Evaluated::Value),
-            StringRef(_) => unreachable(),
-        };
+    };
+}
 
-        let multiply_value = |l: &data::Value, other: &Evaluated<'a>| match other {
-            LiteralRef(r) => l.multiply(&l.clone_by(r)?).map(Evaluated::Value),
-            Literal(r) => l.multiply(&l.clone_by(&r)?).map(Evaluated::Value),
-            ValueRef(r) => l.multiply(r).map(Evaluated::Value),
-            Value(r) => l.multiply(&r).map(Evaluated::Value),
-            StringRef(_) => unreachable(),
-        };
+macro_rules! general_op {
```

Comment:
This is only used for `cast`, so.. how about keeping your original one?

It is because.. the last refactoring would be not a macro, I think we can do refactoring codes again using generic function.
So, this looks like unnecessary margin.

Reply:
Yeah, kinda unneccessary. I agree. Will undo.